stl/inc/format

```diff
@@ -343,72 +337,398 @@ public:
     // _Next_arg_id > 0 means automatic
     // _Next_arg_id == -1 means manual
     constexpr size_t next_arg_id() {
-        if (_Next_arg_id >= 0) {
-            return _Next_arg_id++;
+        if (_Next_arg_id == -1) {
+            throw format_error("Can not switch from manual to automatic indexing");
         }
-        throw format_error("Can not switch from manual to automatic indexing");
+
+        return static_cast<size_t>(_Next_arg_id++);
     }
-    constexpr void check_arg_id(size_t _Id) {
-        (void) _Id;
+
+    constexpr void check_arg_id(const size_t _Id) {
+        if (_STD is_constant_evaluated()) {
+            if (_Id >= _Num_args) {
+                _You_see_this_error_because_arg_id_is_out_of_range();
+            }
+        }
+
         if (_Next_arg_id > 0) {
             throw format_error("Can not switch from automatic to manual indexing");
         }
         _Next_arg_id = -1;
     }
+
+private:
+    basic_string_view<_CharT> _Format_string;
+    size_t _Num_args;
+    // The standard says this is size_t, however we use ptrdiff_t to save some space
+    // by not having to store the indexing mode. Below is a more detailed explanation
+    // of how this works.
+    ptrdiff_t _Next_arg_id = 0;
 };
 
-// TODO: test coverage
+enum class _Basic_format_arg_type : uint8_t {
+    _None,
+    _Int_type,

```

Comment:
I hate that too, at the same time there is the symmetry with `_Int_type` and `_Int_state`. 

Also for those PODs it is difficult to find a good name that is not taken, e.g. `_Bool`

Reply:
I should point out this is a minor annoyance, and I did notice the consistency with `_Meow_state` later in the review. Since I don't have a better suggestion I'll drop this comment.

Reply:
conflicts with the machinery PR, but that's OK, we can fix that up after merging this.